Handle customers (#31)
stakater/jira-service-desk-operator: README.md

$ oc apply -f bundle/manifests
```

### Project
We support the following CRUD operation on project via our Jira Service Desk Operator:
* Create - Creates a new projects with the provided fields
* Update - Updates an existing project with the updated fields
* Delete - Removes and deletes the project 

> An example Custom Resource for a project can be viewed [here](https://github.com/stakater/jira-service-desk-operator/tree/master/examples/project).

Reviewer: ```suggestion
Examples for Project Custom Resource can be found at [here](https://github.com/stakater/jira-service-desk-operator/tree/master/examples/project).
```

#### Limitations:
* We only support creating three types of JSD projects via our operator i.e Business, ServiceDesk, Software. The details and differences between these project types can be viewed [here](https://confluence.atlassian.com/adminjiraserver/jira-applications-and-project-types-overview-938846805.html).

> * Following are the immutable fields and can't be provided while updating a project. 

Reviewer: ```suggestion
* Following are the immutable fields that cannot be updated:
```

    * ProjectTemplateKey
    * ProjectTypeKey
    * leadAccountId 
    * CategoryId 
    * NotificationScheme
    * PermissionScheme 
    * issueSecurityScheme 

    You can read more about these fields on [Jira Service Desk api docs](https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-projects/#api-rest-api-3-project-post).

### Customer:
We support the following CRUD operations on customer via our Jira Service Desk Operator
* Create - Create a new customer and assign the projects mentioned in the CR
* Updates - Update (add/remove) the assigned/associated projects mentioned in the CR

> * Deletes - Remove all the project associations and deletes the customer

Reviewer: Update - Only updates(add/remove) the associated projects mentioned in the CR
Delete -

> An example Custom Resource for Customer can be viewed [here](https://github.com/stakater/jira-service-desk-operator/blob/handle-customers/examples/customer/customer.yaml).

Reviewer: Same as above

#### Limitations:
* Jira Service Desk Operator can access only those customers which are created through it. Customers that are manually created and added in the projects can’t be accessed later with the Jira Service Desk Operator.

> * Each custom resource is associated to a single customer. You can not update **customer name and email**.

Reviewer: > You can not update **customer name and email**.

This should be a separate point

$ oc apply -f bundle/manifests
```

### Create Project

Create the project on jira-service-desk with the following custom resource:

> 

Reviewer: We can create, update and delete a Project using this operator. Kindly, update the documentation and also mention limitations like immutable fields cannot be updated, allowed project types etc.

```yaml
apiVersion: jiraservicedesk.stakater.com/v1alpha1
kind: Project
metadata:
  name: <Name>
  namespace: default
spec:
  name: <ProjectName>
  key: <ProjectKey>
  projectTypeKey: <ProjectTypeKey>
  projectTemplateKey: <ProjectTemplateKey>
  description: <ProjectDescription>
  assigneeType: <ProjectAssigneeType>
  leadAccountId: <ProjectLeadAccountId>

>   url: <ProjectUrl>

Reviewer: Don't add manifest here. Just refer the manifest from `examples/project`
The issue with this is that we'll have to update it in multiple places whenever we update Project CRD

```

### Add/Remove Customers to JSD Project

Create a customer and than add/remove customer into jira-service desk project with the following custom resource

```yaml
apiVersion: jiraservicedesk.stakater.com/v1alpha1
kind: Customer
metadata:
  name: <Name>
spec:
  name: <CustomerName>
  email: <CustomerEmail>
  projects:
    - <AssociatedProjectKey>

> ```

Reviewer: As discussed we were supposed to mention limitations clearly in the documentation. Which is missing, please update this

   - `make bundle`
   - `make packagemanifests`
   

> ## Running Tests
> 
> To run the tests either use `go test` or `ginkgo` in the directory where test_suite.go is located.

Reviewer: ```suggestion
## Running Tests

### Pre-requisites:
1. Create a namespace with the name `test`
2. Create `jira-service-desk-config` secret in test namespace

### To run tests:
Use the following command to run tests:
`make test OPERATOR_NAMESPACE=test USE_EXISTING_CLUSTER=true`
```

   - `make bundle`
   - `make packagemanifests`
   
## Running Tests

### Pre-requisites:
1. Create a namespace with the name `test`
2. Create `jira-service-desk-config` secret in test namespace

### To run tests:
Use the following command to run tests:
`make test OPERATOR_NAMESPACE=test USE_EXISTING_CLUSTER=true`

## Limitations
- Once a customer is created with the Jira Service Desk Operator, it can only be added or removed from the projects. Customers can't be deleted. The Jira Service Desk Api don't provides a support for this.
- Jira Service Desk Operator can access only those customers which are created through it. Customers that are manually created and added in the projects can't be accessed later with the Jira Service Desk Operator.

> - The Api call to remove the customer from the project is in experimental phase at the moment.

Reviewer: These should go in the relevant sections i.e. Project and Customer

> - The Api call to remove the customer from the project is in experimental phase at the moment.

Reviewer: Add a reference for their documentation here
